File: singularity/implementation.java/O7.java
/* Copyright 2018-2019 ComdivByZero
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package o7;

public final class O7 {

public static abstract class ArgsReceiver {
    abstract protected void set(byte[][] args);
}

public static ArgsReceiver argsReceiver = null;

public static final byte BOOL_UNDEF   = (byte)0xFF;
public static final int  INT_UNDEF    = Integer.MIN_VALUE;
public static final long LONG_UNDEF   = Long.MIN_VALUE;
public static final long DOUBLE_UNDEF = 0x7FFFFFFF00000000L;

public static final java.nio.charset.Charset UTF_8
                  = java.nio.charset.Charset.forName("UTF-8");

static int      exitCode  = 0;
static byte[][] args      = new byte[][]{};

private static
    java.lang.ref.SoftReference<java.util.HashMap<java.lang.String, byte[]>>
    stringsCache = new java.lang.ref.SoftReference<java.util.HashMap<java.lang.String, byte[]>>(null);

public static void exit() {
    if (0 != exitCode) {
        java.lang.System.exit(exitCode);
    }
}

public static void asrt(final boolean c) {
    if (!c) {
        throw new java.lang.AssertionError();
    }
}

public static void asrt(final boolean c, final String msg) {
    if (!c) {
        throw new java.lang.AssertionError(msg);
    }
}

public static void caseFail(final int c) {
    throw new java.lang.AssertionError("case fail: " + c);
}

public static boolean inited(final byte b) {
    if (b == BOOL_UNDEF) {
        throw new java.lang.AssertionError("boolean variable is not initialized");
    }
    return b != 0;
}

public static int inited(final int i) {
    if (i == INT_UNDEF) {
        throw new java.lang.AssertionError("int variable is not initialized");
    }
    return i;
}

public static long inited(final long i) {
    if (i == LONG_UNDEF) {
        throw new java.lang.AssertionError("long variable is not initialized");
    }
    return i;
}

public static double inited(final double d) {
    if (java.lang.Double.doubleToRawLongBits(d) == DOUBLE_UNDEF) {
        throw new java.lang.AssertionError("double variable is not initialized");
    }
    return d;
}

public static int ord(final boolean b) {
    final int i;
    if (b) {
        i = 1;
    } else {
        i = 0;
    }
    return i;
}

/* BOOLEAN to integer */
public static int ord(final byte b) {
    inited(b);
    return b;
}

/* SET to integer */
public static int ord(final int i) {
    if (i < 0) {
        throw new java.lang.ArithmeticException("31 is not accepted in SET, when it converted to int");
    }
    return i;
}

public static byte toByte(final int i) {
    if (i < 0 || 0x100 <= i) {
        throw new java.lang.ArithmeticException("int value " + i + " out of byte range");
    }
    return (byte)i;
}

public static byte toByte(final long i) {
    if (i < 0 || 0x100 <= i) {
        throw new java.lang.ArithmeticException("long value " + i + " out of byte range");
    }
    return (byte)i;
}

public static int toInt(final byte b) {
    return b & 0xFF;
}

public static int toInt(final char c) {
    asrt(c <= 255);
    return c;
}

public static int add(final int a, final int b) {
    final int sum;
    //sum = java.lang.Math.addExact(inited(a), inited(b));
    sum = inited(a) + inited(b);
    if (sum == INT_UNDEF) {
        throw new java.lang.ArithmeticException("addition overflow");
    }
    return sum;
}

public static long add(final long a, final long b) {
    final long sum;
    //sum = java.lang.Math.addExact(inited(a), inited(b));
    sum = inited(a) + inited(b);
    if (sum == LONG_UNDEF) {
        throw new java.lang.ArithmeticException("addition overflow");
    }
    return sum;
}

public static int sub(final int a, final int b) {
    final int diff;
    //diff = java.lang.Math.subtractExact(inited(a), inited(b));
    diff = inited(a) - inited(b);
    if (diff == INT_UNDEF) {
        throw new java.lang.ArithmeticException("subtraction overflow");
    }
    return diff;
}

public static long sub(final long a, final int b) {
    final long diff;
    //diff = java.lang.Math.subtractExact(inited(a), inited(b));
    diff = inited(a) - inited(b);
    if (diff == LONG_UNDEF) {
        throw new java.lang.ArithmeticException("subtraction overflow");
    }
    return diff;
}

public static int mul(final int a, final int b) {
    final int prod;
    //prod = java.lang.Math.multiplyExact(inited(a), inited(b));
    prod = inited(a) * inited(b);
    if (prod == INT_UNDEF) {
        throw new java.lang.ArithmeticException("multiply overflow");
    }
    return prod;
}

public static long mul(final long a, final long b) {
    final long prod;
    //prod = java.lang.Math.multiplyExact(inited(a), inited(b));
    prod = inited(a) * inited(b);
    if (prod == LONG_UNDEF) {
        throw new java.lang.ArithmeticException("multiply overflow");
    }
    return prod;
}

public static int div(final int a, final int b) {
    asrt(b >= 0);
    final int mask;

    mask = a >> 31;
    return mask ^ ((mask ^ inited(a)) / b);
}

public static int mod(final int a, final int b) {
    asrt(b >= 0);
    final int mask;

    mask = a >> 31;
    return (b & mask) + (mask ^ ((mask ^ inited(a)) % b));
}

public static int floor(final double d) {
    final double v;
    v = java.lang.Math.floor(inited(d));
    if ((v <= java.lang.Integer.MIN_VALUE) || (java.lang.Integer.MAX_VALUE < v)) {
        throw new java.lang.ArithmeticException("floor overflow");
    }
    return (int)v;
}

public static double scalb(final double d, final int n) {
    asrt(!java.lang.Double.isNaN(d));
    return java.lang.Math.scalb(d, inited(n));
}

public static void scalb(final double d[], final int d_i, final int n) {
    d[d_i] = scalb(d[d_i], n);
}

public static double frexp(final double d, final int[] n, final int n_i) {
    final long bits, mantissa;
    long       divider;
    int        exponent;

    asrt(!java.lang.Double.isNaN(d));

    bits = java.lang.Double.doubleToLongBits(d);

    exponent = (int)((bits >> 52) & 0x07FFL);

    if (exponent == 0) {
        exponent =  1;
        mantissa =  bits & 0x000FFFFFFFFFFFFFL;
    } else {
        mantissa = (bits & 0x000FFFFFFFFFFFFFL) | (1L << 52);
    }

    exponent -= 1075;

    divider = 1;
    while (divider < mantissa) {
        divider  *= 2;
        exponent += 1;
    }
    if (divider != mantissa) {
        divider  /= 2;
        exponent -= 1;
    }

    if (bits < 0) {
       divider = -divider;
    }

    n[n_i] = exponent;
    return ((double)mantissa) / divider;
}

public static double frexp(final double d, final int[] n) {
    return frexp(d, n, 0);
}

public static void frexp(final double d[], final int d_i, final int[] n, final int n_i) {
    d[d_i] = frexp(d[d_i], n, n_i);
}

public static double flt(final int i) {
    return (double)inited(i);
}

public static byte[] bytes(final java.lang.String s) {
    final java.nio.ByteBuffer bb;
    final int len;
    java.util.HashMap<java.lang.String, byte[]> cache;
    byte ba[];

    cache = stringsCache.get();
    if (cache == null) {
        cache = new java.util.HashMap<java.lang.String, byte[]>();
        stringsCache =
            new java.lang.ref.SoftReference<java.util.HashMap<java.lang.String, byte[]>>(cache);
        ba = null;
    } else {
        ba = cache.get(s);
    }
    if (ba == null) {
        bb = UTF_8.encode(s);
        len = bb.limit();
        ba = new byte[len + 1];
        bb.get(ba, 0, len);

        cache.put(s, ba);
    }
    return ba;
}

public static java.lang.String string(final byte[] bytes, final int ofs) {
    int i;
    final java.nio.ByteBuffer buf;

    i = ofs;
    while (bytes[i] != 0) {
         i += 1;
    }
    buf = java.nio.ByteBuffer.wrap(bytes, ofs, i - ofs);
    return UTF_8.decode(buf).toString();
}

public static java.lang.String string(final byte[] bytes) {
    return string(bytes, 0);
}

public static int strcmp(final byte[] s1, final byte[] s2) {
    int i;
    i = 0;
    while ((s1[i] == s2[i]) && (s1[i] != 0)) {
        i += 1;
    }
    return (s1[i] & 0xFF) - (s2[i] & 0xFF);
}

public static int strcmp(final byte[] s1, final byte s2) {
    final int c1, c2;
    int ret;
    c1 = 0xFF & s1[0];
    c2 = 0xFF & s2;
    ret = c1 - c2;
    if (ret == 0 && c1 != 0 && s1[1] != 0) {
        ret = 0xFF & s1[1];
    }
    return ret;
}

public static int strcmp(final byte s1, final byte[] s2) {
    return -strcmp(s2, s1);
}

/* Copy chars */
public static void strcpy(final byte[] d, final byte[] s) {
    int i;

    i = 0;
    while (s[i] != 0) {
        d[i] = s[i];
        i += 1;
    }
    d[i] = 0;
}

public static void strcpy(final byte[] d, final java.lang.String s) {
    final int len;
    final byte[] b;

    b = bytes(s);
    len = b.length;
    for (int i = 0; i < len; i += 1) {
        d[i] = b[i];
    }
    d[len] = 0;
}

public static void copy(java.lang.Object d, java.lang.Object s, int len) {
    java.lang.System.arraycopy(s, 0, d, 0, len);
}

public static void copy(boolean[] d, boolean[] s) {copy(d, s, s.length);}
public static void copy(byte   [] d, byte   [] s) {copy(d, s, s.length);}
public static void copy(int    [] d, int    [] s) {copy(d, s, s.length);}
public static void copy(double [] d, double [] s) {copy(d, s, s.length);}


public static void copy(java.lang.Object d, java.lang.Object s, int ti, java.lang.String type) {
    int len, ilen;
    char ct;
    java.lang.Object item;

    len = java.lang.reflect.Array.getLength(s);
    ct = type.charAt(ti + 1);
    if (ct == '[') {
        for (int i = 0; i < len; i += 1) {
            /* TODO */
            copy(java.lang.reflect.Array.get(d, i),
                 java.lang.reflect.Array.get(s, i), ti + 1, type);
        }
    } else {
        /*TODO*/
        asrt(ct != 'L');

        item = java.lang.reflect.Array.get(s, 0);
        ilen = java.lang.reflect.Array.getLength(item);
        copy(java.lang.reflect.Array.get(d, 0), item, ilen);
        for (int i = 1; i < len; i += 1) {
            copy(java.lang.reflect.Array.get(d, i),
                 java.lang.reflect.Array.get(s, i), ilen);
        }
    }
}

public static void copy(java.lang.Object d, java.lang.Object s) {
    copy(d, s, 1, s.getClass().getName());
}

public static int set(final int low, final int high) {
    asrt(high <= 31);
    asrt(0 <= low && low <= high);
    return (~0 << low) & (~0 >>> (31 - high));
}

public static long lset(final int low, final int high) {
    asrt(high <= 63);
    asrt(0 <= low && low <= high);
    return (~0l << low) & (~0l >>> (63 - high));
}

public static boolean in(final int n, final int set) {
    return (0 <= n) && (n <= 31) && (0 != (set & (1 << n)));
}

public static boolean in(final int n, final long set) {
    return (0 <= n) && (n <= 63) && (0 != (set & (1l << n)));
}

public static int ror(final int n, final int shift) {
    asrt(shift >= 0);
    asrt(n >= 0);
    final int r;
    r = java.lang.Integer.rotateRight(n, shift);
    asrt(r >= 0);
    return r;
}

public static void init(final java.lang.String[] sargs) {
    args = new byte[sargs.length][];
    for (int i = 0; i < sargs.length; i += 1) {
        args[i] = bytes(sargs[i]);
    }
    if (argsReceiver != null) {
        argsReceiver.set(args);
    }
}

public static boolean bit(int addr, int n) {
    asrt(false);
    return false;
}

}

File: singularity/implementation.js/o7.js
/* Copyright 2018-2021 ComdivByZero
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var o7;
(function(o7) { "use strict";

  var utf8Enc, utf8Dec, u8array, toUtf8, utf8Cache, utf8ToStr, proc;

  utf8Cache = [];

  o7.export = {};
  o7.import = o7.export;

  if (typeof process === 'undefined' || !process.exit) {
    proc = {exit : function(code) { if (code != 0) throw code; }};
  } else {
    proc = process;
  }

  function assert(check, msg) {
    if (check) {
      ;
    } else if (msg) {
      throw new Error(msg);
    } else {
      throw new Error("assertion is false");
    }
  }
  o7.assert = assert;

  function indexOut(index, length) {
    return new RangeError("array index - " + index + " out of bounds - 0 .. " + length);
  }

  /* TODO */
  function array() {
    var lens;

    lens = arguments;
    function create(li) {
      var a, len, i;

      len = lens[li];
      a = new Array(len);
      li += 1;
      if (li < lens.length) {
        for (i = 0; i < len; i += 1) {
          a[i] = create(li);
        }
      }
      return a;
    }
    return create(0);
  }
  o7.array = array;

  function sarray() {
    var lens;

    lens = arguments;
    function create(li) {
      var array, len, i;

      len = lens[li];
      array = new Array(len);
      this._ = array;
      this.length = len;
      this.at = function(index) {
        if (0 <= index && index < len) {
          return array[index];
        } else {
          throw indexOut(index, array);
        }
      };
      this.put = function(index, value) {
        if (0 <= index && index < len) {
          array[index] = value;
        } else {
          throw indexOut(index, array);
        }
      };
      this.inc = function(index, val) {
        if (0 <= index && index < len) {
          array[index] = add(array[index], val);
        } else {
          throw indexOut(index, array);
        }
      };
      this.incl = function(index, val) {
        if (0 <= index && index < len) {
          array[index] |= incl(val);
        } else {
          throw indexOut(index, array);
        }
      };
      this.excl = function(index, val) {
        if (0 <= index && index < len) {
          array[index] &= excl(val);
        } else {
          throw indexOut(index, array);
        }
      };
      li += 1;
      if (li < lens.length) {
        for (i = 0; i < len; i += 1) {
          array[i] = new create(li);
        }
      }
    }
    return new create(0);
  }
  o7.sarray = sarray;

  o7.at = function(array, index) {
    if (0 <= index && index < array.length) {
      return array[index];
    } else {
      throw indexOut(index, array.length);
    }
  };

  o7.put = function(array, index, value) {
    if (0 <= index && index < array.length) {
      array[index] = value;
    } else {
      throw indexOut(index, array.length);
    }
  };

  function ind(index, length) {
    if (0 <= index && index < length) {
      return value;
    } else {
      throw indexOut(index, length);
    }
  }
  o7.ind = ind;

  function getjsa(o7array) {
    var ja;
    if (o7array._) {
      ja = o7array._;
    } else {
      ja = o7array;
    }
    return ja;
  }
  o7.getjsa = getjsa;

  o7.caseFail = function(val) {
    throw new RangeError("Unexpected value in case = " + val);
  };

  o7.cti = function(char) {
    return char.charCodeAt(0);
  };

  o7.itc = function(int) {
    if (0 <= int && int < 0x100) {
      return int;
    } else {
      throw new RangeError("Char range overflow during cast from " + int);
    }
  };

  o7.bti = function(bool) {
    var i;
    if (bool) {
      i = 1;
    } else {
      i = 0;
    }
    return i;
  };

  o7.sti = function(bitset) {
    if (0 <= bitset && bitset < 0x80000000) {
      return bitset;
    } else {
      throw new RangeError("Set " + bitset + " can not be converted to integer");
    }
  };

  o7.itb = function(int) {
    if (0 <= int && int < 0x100) {
      return int;
    } else {
      throw new RangeError("Byte range is overflowed during cast from " + int);
    }
  };

  o7.floor = function(double) {
    var v;
    v = Math.floor(double);
    if ((-0x80000000 < v) && (v < 0x80000000)) {
      return v;
    } else {
      throw new RangeError("floor overflow " + v);
    }
  };

  o7.flt = function(int) {
    /* TODO */
    return int;
  };

  o7.scalb = function(double, int) {
    /* TODO */
    return double * Math.pow(2, int);
  };

  o7.frexp = function(d, n, n_i) {
    /* TODO */
    var abs, exp, x;
    n = getjsa(n);
    if (d !== 0.0) {
      abs = Math.abs(d);
      exp = Math.max(-1023, Math.floor(Math.log(abs) * Math.LOG2E) + 1);
      x = abs * Math.pow(2, -exp);

      while (x < 1.0) {
        x   *= 2.0;
        exp -= 1;
      }
      while (x >= 2.0) {
        x   /= 2.0;
        exp += 1;
      }
      if (d < 0.0) {
        x = -x;
      }
      n[n_i] = exp;
    } else {
      x      = 0.0;
      n[n_i] = 0.0;
    }
    return x;
  }

  o7.in = function(n, st) {
    return (0 <= n) && (n <= 31) && (0 != (st & (1 << n)));
  };

  if (typeof Uint8Array !== 'undefined') {
    u8array = function(array) {
      return new Uint8Array(array);
    }
  } else {
    u8array = function(array) {
      return array;
    };
  }

  u8array = function(array) {
    function create() {
      var len;

      len = array.length;
      this.length = len;
      this._ = array;
      this.at = function(index) {
        if (0 <= index && index < len) {
          return array[index];
        } else {
          throw indexOut(index, array);
        }
      };
      this.put = function(index, val) {
        if (0 <= index && index < len) {
          array[index] = val;
        } else {
          throw indexOut(index, array);
        }
      };
      this.inc = function(index, val) {
        if (0 <= index && index < len) {
          val += array[index];
          if (0 <= val && val <= 0xFF) {
            array[index] = val;
          }
        } else {
          throw indexOut(index, array);
        }
      };
    }
    return new create();
  };

  function arrayUtf8ToStr(bytes) {
    var str, buf, i, len, ch, ch1, ch2, ch3, ok;

    buf = [];
    len = bytes.length;
    i = 0;
    ok = true;
    while (i < len && bytes[i] != 0) {
      ch = bytes[i];
      i += 1;
      if (ch < 0x80) {
        buf.push(String.fromCharCode(ch));
      } else if (ch < 0xC0) {
        ok = false;
      } else if (ch < 0xE0) {
        if (i < len) {
          ch1 = bytes[i];
          i += 1;
          if ((ch1 >> 6) == 2) {
            buf.push(String.fromCharCode(((ch & 0x1F) << 6) | (ch1 & 0x3F)));
          } else {
            ok = false;
          }
        } else {
          ok = false;
        }
      } else if (ch < 0xF0) {
        if (i + 1 < len) {
          ch1 = bytes[i];
          ch2 = bytes[i + 1];
          i += 2;
          if (((ch1 >> 6) == 2) && ((ch2 >> 6) == 2)) {
            buf.push(String.fromCharCode(((ch & 0xF) << 12) | ((ch1 & 0x3F) << 6) | (ch2 & 0x3F)));
          } else {
            ok = false;
          }
        } else {
          ok = false;
        }
      } else {
        if (i + 2 < len) {
          ch1 = bytes[i];
          ch2 = bytes[i + 1];
          ch3 = bytes[i + 2];
          i += 3;
          if (((ch1 >> 6) == 2) && ((ch2 >> 6) == 2) && ((ch3 >> 6) == 2)) {
            buf.push(String.fromCodePoint(((ch & 0x7) << 18) | ((ch1 & 0x3F) << 12) | ((ch2 & 0x3F) << 6) | (ch3 & 0x3F)));
          } else {
            ok = false;
          }
        } else {
          ok = false;
        }
      }
    }
    if (ok) {
      str = buf.join('');
    } else {
      str = null;
    }
    return str;
  };

  if (typeof TextDecoder !== 'undefined') {
    utf8Enc = new TextEncoder('utf-8');
    utf8Dec = new TextDecoder('utf-8');

    if (utf8Enc.encode("!").push) {
      toUtf8 = function(str) {
        var a;
        a = utf8Enc.encode(str);
        a.push(0);
        return u8array(a);
      };
    } else {
      toUtf8 = function(str) {
        var a, b;
        a = utf8Enc.encode(str);
        b = new Uint8Array(a.length + 1);
        b.set(a, 0);
        b[a.length] = 0;
        return u8array(b);
      };
    }

    utf8ToStr = function(bytes) {
      var str;
      if (bytes instanceof Uint8Array) {
        str = utf8Dec.decode(bytes);
      } else {
        str = arrayUtf8ToStr(bytes);
      }
      return str;
    };
  } else {
    /* str must be correct utf16 string */
    toUtf8 = function(str) {
      var bytes, si, ch, len;
      bytes = [];
      si = 0;
      len = str.length;
      while (si < len) {
        ch = str.charCodeAt(si);
        if (ch < 0x80) {
          bytes.push(ch);
        } else if (ch < 0x800) {
          bytes.push((ch >> 6) | 0xC0,
                     (ch & 0x3F) | 0x80);
        } else if ((ch & 0xFC00) == 0xD800) {
          si += 1;
          ch = 0x10000 | ((ch & 0x3FF) << 10) | (str.charCodeAt(si) & 0x3FF);
          bytes.push((ch >> 18) | 0xF0,
                     ((ch >> 12) & 0x3F) | 0x80,
                     ((ch >> 6 ) & 0x3F) | 0x80,
                     (ch & 0x3F) | 0x80);
        } else {
          bytes.push((ch >> 12) | 0xE0,
                     ((ch >> 6) & 0x3F) | 0x80,
                     (ch & 0x3F) | 0x80);
        }
        si += 1;
      }
      bytes.push(0x0);
      return u8array(bytes);
    };

    utf8ToStr = arrayUtf8ToStr;
  }
  o7.utf8ToStr = utf8ToStr;

  o7.toUtf8 = function(str) {
    var utf;
    utf = utf8Cache[str];
    if (!utf) {
        utf = toUtf8(str);
        utf8Cache[str] = utf;
    }
    return utf;
  };

  o7.utf8ByOfsToStr = function(bytes, ofs) {
    bytes = getjsa(bytes);
    if (ofs > 0) {
      bytes = bytes.slice(ofs);
    }
    return utf8ToStr(bytes);
  }

  /* str must be correct 7bit ASCII string */
  o7.toAscii = function(str) {
    var bytes, len, i;
    len = str.length;
    bytes = new Uint8Array(len);
    for (i = 0; i < len; i += 1) {
      /* assert str.charCodeAt(i) < 0x80 */
      bytes[i] = str.charCodeAt(i);
    }
    return bytes;
  };


  o7.extend = function(ext, base) {
    function proto() {}
    proto.prototype = base.prototype;

    ext.prototype = new proto();
    ext.base = base;
    return ext;
  };

  function add(a, b) {
    var r;
    r = a + b;
    if (-0x80000000 < r && r < 0x80000000) {
      return r;
    } else {
      throw new RangeError("integer overflow in " + a + " + " + b + " = " + r);
    }
  }
  o7.add = add;

  o7.sub = function(a, b) {
    var r;
    r = a - b;
    if (-0x80000000 < r && r < 0x80000000) {
      return r;
    } else {
      throw new RangeError("integer overflow in " + a + " - " + b + " = " + r);
    }
  };

  o7.mul = function(a, b) {
    var r;
    r = a * b;
    if (-0x80000000 < r && r < 0x80000000) {
      return r;
    } else {
      throw new RangeError("integer overflow in " + a + " * " + b + " = " + r);
    }
  };

  o7.div = function(a, b) {
    var mask;
    if (b > 0) {
      mask = a >> 31;
      return mask ^ ((mask ^ a) / b);
    } else {
      throw new RangeError("Integer divider can't be < 1");
    }
  };

  o7.mod = function(a, b) {
    var mask;
    if (b > 0) {
      mask = a >> 31;
      return (b & mask) + (mask ^ ((mask ^ a) % b));
    } else {
      throw new RangeError("Integer divider can't be < 1");
    }
  };

  function fadd(a, b) {
    var s;
    s = a + b;
    if (isFinite(s)) {
      return s;
    } else {
      /* TODO */
      throw new RangeError("Fraction out of range in " + a + " + " + b + " = " + s);
    }
  }
  o7.fadd = fadd;

  function fsub(a, b) {
    var s;
    s = a - b;
    if (isFinite(s)) {
      return s;
    } else {
      /* TODO */
      throw new RangeError("Fraction out of range in " + a + " - " + b + " = " + s);
    }
  }
  o7.fsub = fsub;

  function fmul(a, b) {
    var s;
    s = a * b;
    if (isFinite(s)) {
      return s;
    } else {
      /* TODO */
      throw new RangeError("Fraction out of range in " + a + " * " + b + " = " + s);
    }
  }
  o7.fmul = fmul;

  function fdiv(a, b) {
    var s;
    s = a / b;
    if (isFinite(s)) {
      return s;
    } else {
      /* TODO */
      throw new RangeError("Fraction out of range in " + a + " / " + b + " = " + s);
    }
  }
  o7.fdiv = fdiv;

  o7.set = function(low, high) {
    if (high > 31) {
      throw new RangeError("high limit = " + high + " > 31");
    } else if (low < 0) {
      throw new RangeError("low limit = " + low + " < 0");
    } else if (low > high) {
      throw new RangeError("low limit = " + low + " > " + high + " - high limit");
    } else {
      return (~0 << low) & (~0 >>> (31 - high));
    }
  };

  function setRangeError(val) {
    return new RangeError("set item = " + val + " out of range 0 .. 31");
  }

  function incl(val) {
    if (0 <= val && val <= 31) {
      return 1 << val;
    } else {
      throw setRangeError(val);
    }
  }
  o7.incl = incl;

  function excl(val) {
    if (0 <= val && val <= 31) {
      return ~(1 << val);
    } else {
      throw setRangeError(val);
    }
  }
  o7.excl = excl;

  o7.ror = function(n, shift) {
    assert(n     >= 0);
    assert(shift >= 0);
    shift &= 31;
    n = (n >>> shift) | (n << (32 - shift));
    assert(n     >= 0);
    return n;
  }

  function inited(val) {
    if (isFinite(val)) {
      return val;
    } else {
      throw new RangeError("Uninitialized value");
    }
  }
  o7.inited = inited;

  o7.cmp = function(a, b) {
    var d;
    d = a - b;
    if (isFinite(d)) {
      ;
    } else if (inited(a) < inited(b)) {
      d = -1;
    } else if (a > b) {
      d = 1;
    } else {
      d = 0;
    }
    return d
  }

  o7.strcmp = function(s1, s2) {
    var i;
    i = 0;
    s1 = getjsa(s1);
    s2 = getjsa(s2);
    while ((s1[i] == s2[i]) && (s1[i] != 0)) {
      i += 1;
    }
    return s1[i] - s2[i];
  };

  function strchcmp(s1, c2) {
    var c1, ret;

    s1 = getjsa(s1);
    c1 = s1[0];
    ret = c1 - c2;
    if (ret == 0 && c1 != 0 && s1.length > 1 && s1[1] != 0) {
        ret = s1[1];
    }
    return ret;
  }
  o7.strchcmp = strchcmp;

  o7.chstrcmp = function(c1, s2) {
    return -strchcmp(s2, c1);
  };

  /* Copy chars */
  o7.strcpy = function(d, s) {
    var len, i;

    d = getjsa(d);
    s = getjsa(s);

    len = s.length;
    assert(d.length >= len);
    for (i = 0; i < len; i += 1) {
      d[i] = s[i];
    }
    assert(d[len - 1] == 0);
  };

  function copy(d, s) {
    var i, len;
    d = getjsa(d);
    s = getjsa(s);
    len = d.length;
    if ((s[0] instanceof Object) && s[0].length) {
      for (i = 0; i < len; i += 1) {
        copy(d[i], s[i]);
      }
    } else {
      for (i = 0; i < len; i += 1) {
        d[i] = s[i];
      }
    }
  }
  o7.copy = copy;

  o7.exit_code = 0;
  o7.main = function(main) {
    main();
    if (o7.exit_code != 0) {
      proc.exit(o7.exit_code);
    }
  };

}) (o7 || (o7 = {}));
